src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -2069,6 +2219,183 @@ void M(bool b)
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "Test2").WithArguments("Test2<T1, T2>(T2, delegate*<T1, T2>)").WithLocation(5, 5),
+                // (6,38): error CS1660: Cannot convert lambda expression to type 'delegate*<string, int>' because it is not a delegate type
+                //     Test1<string, int>(string.Empty, v => 0);
+                Diagnostic(ErrorCode.ERR_AnonMethToNonDel, "v => 0").WithArguments("lambda expression", "delegate*<string, int>").WithLocation(6, 38),
+                // (7,27): error CS1660: Cannot convert lambda expression to type 'delegate*<string, int>' because it is not a delegate type
+                //     Test2<string, int>(0, v => 0);
+                Diagnostic(ErrorCode.ERR_AnonMethToNonDel, "v => 0").WithArguments("lambda expression", "delegate*<string, int>").WithLocation(7, 27)
+            );
+        }
+
+        [Fact, WorkItem(50096, "https://github.com/dotnet/roslyn/issues/50096")]
+        public void AddressOfInference_OnDelegateType()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+using System;
+unsafe
+{
+    Test1(string.Empty, &converter);
+    Test2(0, &converter);
+    Test1<string, int>(string.Empty, &converter);
+    Test2<string, int>(0, &converter);
+
+    static int converter(string v) => 0;
+    static void Test1<T1, T2>(T1 t1, Func<T1, T2> func) {}
+    static void Test2<T1, T2>(T2 t2, Func<T1, T2> func) {}
+}
+", options: TestOptions.UnsafeReleaseExe);
+
+            comp.VerifyDiagnostics(
+                // (5,5): error CS0411: The type arguments for method 'Test1<T1, T2>(T1, Func<T1, T2>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //     Test1(string.Empty, &converter);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "Test1").WithArguments("Test1<T1, T2>(T1, System.Func<T1, T2>)").WithLocation(5, 5),
+                // (6,5): error CS0411: The type arguments for method 'Test2<T1, T2>(T2, Func<T1, T2>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //     Test2(0, &converter);
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "Test2").WithArguments("Test2<T1, T2>(T2, System.Func<T1, T2>)").WithLocation(6, 5),
+                // (7,38): error CS1503: Argument 2: cannot convert from '&method group' to 'Func<string, int>'
+                //     Test1<string, int>(string.Empty, &converter);
+                Diagnostic(ErrorCode.ERR_BadArgType, "&converter").WithArguments("2", "&method group", "System.Func<string, int>").WithLocation(7, 38),
+                // (8,27): error CS1503: Argument 2: cannot convert from '&method group' to 'Func<string, int>'
+                //     Test2<string, int>(0, &converter);
+                Diagnostic(ErrorCode.ERR_BadArgType, "&converter").WithArguments("2", "&method group", "System.Func<string, int>").WithLocation(8, 27)
+            );
+        }
```

Comment:
>} [](start = 8, length = 1)

It looks like the change of "If Ui is not known to be a reference type then an exact inference is made" to "If Ui is not known to be a reference type or a function pointer type then an exact inference is made" in the specification enables the inference through arrays and generic types. Is this intentional? Are we testing that?

Reply:
The spec has been updated to address this.

---
In reply to: [558722993](https://github.com/dotnet/roslyn/pull/50249#discussion_r558722993) [](ancestors = 558722993)